cores/esp8266/Esp.cpp

```diff
@@ -140,6 +140,168 @@ uint64_t EspClass::deepSleepMax()
+                dest = dest->timer_next;
+            }
+        }
+        DEBUG_SERIAL.printf("=============\n");
+        for (os_timer_t* timer_node = timer_root; nullptr != timer_node; timer_node = timer_node->timer_next) {
+            DEBUG_SERIAL.printf("timer_address = %p\n", timer_node);
+            DEBUG_SERIAL.printf("timer_expire  = %u\n", timer_node->timer_expire);
+            DEBUG_SERIAL.printf("timer_period  = %u\n", timer_node->timer_period);
+            DEBUG_SERIAL.printf("timer_func    = %p\n", timer_node->timer_func);
+            DEBUG_SERIAL.printf("timer_next    = %p\n", timer_node->timer_next);
+            if (timer_node->timer_next) DEBUG_SERIAL.printf("=============\n");
+        }
+        DEBUG_SERIAL.printf("=============\n");
+        DEBUG_SERIAL.flush();
+        while (timer_root) {
+            auto next = timer_root->timer_next;
+            delete timer_root;
+            timer_root = next;
+        }
+    }
+}
+#endif
+
+bool EspClass::forcedLightSleepBegin(uint32_t duration_us, fpm_wakeup_cb wakeupCb)
+{
+    // Setting duration to 0xFFFFFFF, it disconnects the RTC timer
+    if (!duration_us || duration_us > 0xFFFFFFF) {
+        duration_us = 0xFFFFFFF;
+    }
+    wifi_fpm_close();
+    saved_sleep_type = wifi_fpm_get_sleep_type();
+    wifi_set_opmode(NULL_MODE);
+    wifi_fpm_set_sleep_type(LIGHT_SLEEP_T);
+    wifi_fpm_open();
+    saved_wakeupCb = wakeupCb;
+    wifi_fpm_set_wakeup_cb([]() {
+        if (saved_wakeupCb) {
+            saved_wakeupCb();
+            saved_wakeupCb = nullptr;
+        }
+        esp_schedule();
+        });
+#ifdef DEBUG_SERIAL
+    walk_timer_list();
+#endif
+    {
+        esp8266::InterruptLock lock;
+        saved_timer_list = timer_list;
+        timer_list = nullptr;
+    }
+    return wifi_fpm_do_sleep(duration_us) == 0;
+}
+
+void EspClass::forcedLightSleepEnd(bool cancel)
+{
+    if (!cancel) {
+        // SDK turns on forced light sleep in idle task
+#ifdef HAVE_ESP_SUSPEND
+        esp_suspend();
+#else
+        esp_yield();
+#endif
+    }
+#ifdef DEBUG_SERIAL
+    walk_timer_list();
+#endif
+    {
+        esp8266::InterruptLock lock;
+        timer_list = saved_timer_list;
+    }
+    saved_wakeupCb = nullptr;
+    wifi_fpm_close();
+    wifi_fpm_set_sleep_type(saved_sleep_type);
+    saved_sleep_type = NONE_SLEEP_T;
+    if (cancel) {
+        // let the SDK catch up in idle task
+#ifdef HAVE_ESP_SUSPEND
+        esp_delay(10);
+#else
+        delay(10);
+#endif
+    }
+}
+
+void EspClass::autoModemSleep() {
+    wifi_fpm_close();
+    saved_sleep_type = wifi_get_sleep_type();
+    wifi_set_sleep_type(MODEM_SLEEP_T);
+}
+
+void EspClass::autoLightSleep() {
+    wifi_fpm_close();
+    saved_sleep_type = wifi_get_sleep_type();
+    wifi_set_sleep_type(LIGHT_SLEEP_T);
+}
+
+void EspClass::autoSleepOff() {
+    wifi_set_sleep_type(saved_sleep_type);
+    saved_sleep_type = NONE_SLEEP_T;
+}
+
```

Comment:
Please elaborate further.

Reply:
@mcspr I've researched based on your open-ended question and I can honestly say that until now, we haven't measured the effects of the SDK version 3 calls,  that `setSleepMode` and friends provide hints for. Plus I honestly have no opinion.
> Why sync this up with forced sleep though, since both do different things either when WiFi is ON or OFF?

I don't understand why this is a question and not an expression of opinion, please explain what you prefer and be specific.

Reply:
The APIs in question are `wifi_.et_sleep_level()` and `wifi_.et_sleep_interval()` (regex notation) . They were updated in this repo by 8ef21ca3aeff76d1265a72ac106999ed9c631a7e on 2018-10-09, which speaking in release numbers places them in core version 2.5.0 and onward. I don't know enough about SDK version numbering to make any sense of v2.2 in the commit message, the online SDK docs, but `NONOSDK3V0` in the code's `#ifdef`s, but there was #5919 .
I don't know that there is a need for all this mixing and matching via preprocessor defines years back into the past going to core milestone 3.0.0, so perhaps those code variants could be cleaned up, I'm sure @d-a-v  is the right person to get involved in deciding that.

Also, it's getting out of control for a single PR. I'm glad for the hints, we are investigation the effects on power from altering sleep level and interval for the ESP class functions proposed here, but any discussion of further integration with ESP8266WiFi should probably occur in a separate thread dedicated to that library.

Reply:
Perhaps... #7956 is still on hold though

This is not the question about the effects though, but the API usage and why it is used that way. Right now there is a single value storing both WiFi sleep and FPM sleep, which may be saved / restored at different points throughout the ESP method calls. Why set / get both interchangeably? Shouldn't there be a separate value for fpm and another one for modem sleep?